[3.3.5] Scripts/Northrend: Trial of the Champion

Wilcalaf opened the pull request:
I adapted the code from the [`3.3.5-trial_of_the_champion`](https://github.com/TrinityCore/TrinityCore/tree/3.3.5-trial_of_the_champion) branch to compile with newer 3.3.5
Tested multiple times, no crashes found yet.

I didn't do very much. just changed a few lines.
it works with the DB changes committed on the original branch here: https://github.com/TrinityCore/TrinityCore/commit/4ea9d7d6a7d8da0edae3c3715bcb0de727299e10

in https://github.com/TrinityCore/TrinityCore/issues/383 tkrokli advised me to make a pull request to the [`3.3.5-trial_of_the_champion`](https://github.com/TrinityCore/TrinityCore/tree/3.3.5-trial_of_the_champion) branch, but I had trouble doing that so I prefer to do it here.

Sorry in advance if there is some wrong stuff. I am very new to Trinity core, I would like some help to fix the issues that i found...

**Changes proposed:**

-  Adaptation from the [`3.3.5-trial_of_the_champion`](https://github.com/TrinityCore/TrinityCore/tree/3.3.5-trial_of_the_champion) branch to [`3.3.5`](https://github.com/TrinityCore/TrinityCore/tree/3.3.5).

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)


**Tests performed:** (Does it build, tested in-game, etc.)

it compiles and looks working and stable but has the issues mentioned above (maybe more)

**Known issues and TODO list:** (add/remove lines as needed)


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Treeston commented:
@Wilcalaf are you planning to work on this?

Treeston commented:
For `CastCustomSpell` -> `CastSpell` take a look at #21123

Wilcalaf commented:
@Treeston Im trying to. but as i said im pretty new to Trinity core. i was able to adapt and fix some issues by myself (like 2 second boss dying instead of becoming friendly and speaking to you) atm im stucked a little with the issues i addresses. if i can fix them i learn. im having a blast dealing with this to be honest hahaha

Gooyeth commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
Copyright 😋

tkrokli replied:
Yeah, I should have updated the Copyright end year in the TC branch `3.3.5-trial_of_the_champion`.
Thanks for the reminder. ^^

Treeston commented:
Sure, no danger in learning on a script nobody else wants to deal with anyway 😆 

(get on [IRC](https://qchat.rizon.net/?channels=trinity) if you want real-time help)

Jildor commented:
in my opinion, this is not the best way to do it.
@tkrokli or someone can update 3.3.5-trial_of_the_champion branch with latest 3.3.5 changes?

Once it is updated, @Wilcalaf you can create a pull request to correct compile errors and warnings (remember target 3.3.5-trial_of_the_champion)

and then you can open little pull request for update the code bit a bit ;)

Treeston commented:
What's the benefit? None of the code in that branch has ever been properly reviewed either.

Jildor commented:
Avoid past mistakes (see the previous PR about this)

Wilcalaf commented:
@Jildor , @Treeston thank you for your replies :). i will try to fix whatever you think is not appropiate or outdated. i started dealing with this for fun and saw that it actually worked very good in game thats why i decided to share it. i think its a really good opportunity to learn by fixing this based on your reviews. so basically ill do whatever you guys tell me to hahaha (provided that i have the required skill ofc)

Wilcalaf committed:
Test.

tkrokli commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -684,12 +1158,114 @@ class spell_paletress_summon_memory : public SpellScriptLoader
+                return true;
+            }
+
+            void HandleScript(AuraEffect* aurEff, DamageInfo& dmgInfo, uint32& absorbAmount)
```
Travis (travis-ci) complains here because the parameter `aurEff` is not used in the function directly below:
```cpp
src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:1177:43:
 fatal error: unused parameter 'aurEff' [-Wunused-parameter]
            void HandleScript(AuraEffect* aurEff, DamageInfo& dmgInfo, uint32& absorbAmount)
                                          ^
```
You can fix it by adding comment separators around the parameter like this: `/*aurEff*/`

(BTW, travis-ci stops looking for these when it has found one, so there could be more further down.)

Wilcalaf replied:
Thank you - i noticed this already. im doing some more changes and commiting more updates soon

tkrokli commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -17,64 +17,118 @@
 SDName: Argent Challenge Encounter.
-SD%Complete: 50 %
-SDComment: AI for Argent Soldiers are not implemented. AI from bosses need more improvements.
+SD%Complete: 100%
```
We may want to get rid of the ScriptData (SD) comment section at a later date, when we are reasonably certain we have replaced the MaNGOS code remnants in the script. Comments are welcome on this part, since I am not certain about the details myself.

Treeston commented:
Get rid of mangos stuff since this is a rewrite, and we don't care about travis until this is merge ready.

Commit:
ToC Almost working -
Adapted the code to compile with latest Trinity core rev.
Tested multiple times, no crashes found yet.

Known issues: i disabled Palettres shield while memory is active, since i can not make the event that removes it happen . need help with that (old one was crashing server)
Palettress spell reflect is not actually reflecting damage. the old function was CastCustomSpell looks like it doesnt exist anymore. looks like you have to use CastSpell with some modifiers but i still havent figured out how. also help needed

Foereaper commented:
I actually started porting this script myself and have it working currently, what I did for Paletress' spell reflect was this:

```cpp
    void HandleScript(AuraEffect* aurEff, DamageInfo& dmgInfo, uint32& absorbAmount)
        {
            // Reflecting 25% of absorbed damage back to attacker
            if (dmgInfo.GetAttacker() == GetTarget())
                return;

            int32 bp = CalculatePct(absorbAmount, GetSpellInfo()->Effects[EFFECT_2].CalcValue());

            GetTarget()->CastSpell(dmgInfo.GetAttacker(), SPELL_SHIELD_REFLECT, CastSpellExtraArgs(aurEff).AddSpellBP0(bp));
        }
```

I don't know how correct this is, but seems to work! Let me know what other issues you have specifically and I'll post what I have done so far :)

Treeston commented:
Looks correct to me :+1:

Wilcalaf commented:
@Foereaper Hey Thanks for help!
So if whats left from the issues i found is 
Make Patetress remove the spell shield after killing the memory npc (im trying to fix it myself now),
and make Eadrec award his part for the dungeon achievement.
a new issue that i found is that the Dark knight flies too slow.
thats the in game issues i found ... now for the code issues we have our guys here that will help us telling is what is acceptable and what now :)

Foereaper commented:
Hmm, the shield removal seems to be working correctly on my end. Want me to upload the file I have so you can compare?

The dark knight flight speed I have fixed already, not entirely sure where I did that exactly but I believe I did that in the movement speed of the gryphon mount in the database. I will double check.

Wilcalaf commented:
@Foereaper Yes that would help, Thank you!

Foereaper commented:
https://gist.github.com/Foereaper/5c5a3fca0f582fc10b9082bb1494ca54

Wilcalaf commented:
@Foereaper On what trinity core Rev this is workng for you?? latest? Unit* summoner = me->ToTempSummon()->GetSummoner() This line used to crash for me that what im trying to adapt.

Foereaper commented:
The achievement part I have not had any time to look into myself yet

Wilcalaf commented:
@Foereaper , damage = me->GetHealth() - 1; This is also working for you? on my tests the without any comands you  were able to kill Paletress (or the other one) sometimes,  and not triggering the rest of the event.

Foereaper commented:
I have had it running on my test realm with a group of people playing for quite some time, no reports of them dying and not moving to the next event in the chain

Wilcalaf commented:
@Foereaper I just tested with your file as is. it didnt crash - altough the issue of the Boss Dying and not triggering the event is also here.
Ill continue testing and maybe apply my fix to this to your base file.
Im comparing both files there are some differences, but im still not sure why mine crashed yet.
Anyway thank you so much for your help

Foereaper commented:
No problem, let me know if there is anything else you need, I did some other fixes to some of the other files as well so I might have some more fixes for you when you get to it :)

Treeston commented:
@Wilcalaf which line are we talking about in your branch?

Foereaper commented:
@Treeston https://github.com/Wilcalaf/TrinityCore/blob/672291411cdd00ee4822a89bbf09dfbf27a92589/src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp#L292

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
+
+            void JustDied(Unit* /*killer*/) override
+            {
+                /*if (Unit* summoner = me->ToTempSummon()->GetSummoner())
```
Use `SummonedCreatureDies` on Confessor's AI.

Wilcalaf replied:
@Treeston Done - Thanks for advice :)

Wilcalaf commented:
@Treeston What Foreaper linked is what i did to stop any of thoose bosses from dying sometimes and not executing the rest of the event.

This line (i commented it out) Was making mine crash. i tried to find another way as you can see a little bit lower in the code

https://github.com/Wilcalaf/TrinityCore/blob/05011d80f4bdadc60bd3389381ae41964c1b2561/src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp#L899

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
-                        DoCast(target, SPELL_SHADOWS_PAST);
+                    switch (eventId)
+                    {
+                        case EVENT_ENTER_AGGRESSIVE:
```
I don't really like this event, because it requires `UpdateVictim` to be after the eventmap update. Can we move it to a timer on Palatress' AI instead?

Wilcalaf replied:
@Treeston What would be the right way to do this? using the same event already on the Paletress AI or adding another event on the AI? if so would this imply having 2 different whiles on the UpdateAI of the Paletress? or i got all wrong? maybe you remember a script were i can look for reference?

Treeston replied:
Huh? Just add a second event to Palatress' `enum`, and a `case` for it on Palatress' AI, and schedule that event however many seconds after the initial summon.

Wilcalaf replied:
I think i got it now. thank you again!

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
+            {
+                Initialize();
+				instance = creature->GetInstanceScript();
+                me->SetReactState(REACT_PASSIVE);
```
Just passive, but still attackable? Should there be immunity flags here?

Foereaper replied:
iirc, the "image" would spawn, and would be attackable while "spawning", but would not react until it entered the aggressive event, so this is correct

Foereaper replied:
Actually, reviewing some old footage of the fight, the memory would be attackable, and would receive debuffs, however it would not take any damage until it entered the aggressive event. So yeah, should not take any damage while in the passive state.

Wilcalaf replied:
Ummm isnt all the party under Confess effect here? (stunned)

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
-        }
+            npc_fountain_of_lightAI(Creature* creature) : ScriptedAI(creature)
+            {
+                me->SetReactState(REACT_PASSIVE);
```
Consider using `NullCreatureAI` here.

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
+                {
+                    _shielded = true;
+                    damage = me->GetHealth() - 1;
+                    me->AttackStop();
```
consider `REACT_PASSIVE`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
+                            break;
+                        case EVENT_RENEW:
+                            me->InterruptNonMeleeSpells(true);
+                            if ((rand32() % 2) > 0 && me->GetHealthPct() < 100)
```
`(rand32() % 2) > 0` is really ugly

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,512 +158,931 @@ enum Spells
+                    return;
+                }
+
+                if (!_memorySummoned && !HealthAbovePct(25))
```
`HealthAbovePctDamaged`, maybe?

Treeston replied:
also, consider capping damage to keep her at just under 25% HP here - otherwise you can still overkill to skip this phase

Wilcalaf replied:
@Treeston  HealthAbovePctDamaged? I cant find this function anywhere

Treeston replied:
`HealthBelowPctDamaged`, my bad.

Do Something like
```c++
if (!_memorySummoned && me->HealthBelowPctDamaged(25, damage))
{
    damage = me->GetHealth() - me->CountPctFromMaxHealth(25);
    /* summon etc */
```

Treeston commented:
see my comment on palatress' damagetaken

Commit:
unused variable commented out until a fix

Foereaper commented:
Oh yeah, another issue I found with the original script, when you knock the champions off their mounts, you cannot trample them with your own mount, and they walk straight back to a new mount

Wilcalaf committed:
Removed the Mangos Comments as requested.
Based the argent_challenge on Foereaper version.

Wilcalaf committed:
Cleaned up a little

Wilcalaf commented:
Guys i updated the file boss_argent_challenge.cpp to be based on what Foereaper showed - so im applying the changes reviewed by @Treeston there (because most of them are also applicable there too)
ill commit more soon

Wilcalaf committed:
Moved Aggressive event from Memory AI to Paletress AI
Used roll_chance_i
flags on memory before he spawns

Wilcalaf committed:
Attempt to fix the complain from the automated check.

tkrokli commented:
Not sure which editor you are using, but you need to set it to auto convert tabs to 4 spaces.

Wilcalaf committed:
Spaces to Tabs

Wilcalaf commented:
@tkrokli Omg i didnt read it right and did the opposite lol, sorry ill fix this

Gooyeth commented:
@Wilcalaf: Nothing happens 😋

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
revert this change

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -1,695 +1,1265 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
revert this change

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
revert this change

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
revert this change

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/instance_trial_of_the_champion.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
revert this change

Aokromes commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
revert this change

Wilcalaf committed:
Tab to space
and copyright notes

tkrokli commented:
Thank you, I found it funny seeing that commit. By the way, next time you visit `trial_of_the_champion.h`, make sure you add a missing New Line (or `<Enter>`/Line Feed, if you are unfamiliar with NL) at the very end, it got removed in your last update (maybe overzealous tab removal, I don't know).

tkrokli commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.cpp:
```diff
@@ -15,56 +15,262 @@
+
+#define GOSSIP_START_EVENT1     "I am ready."
+#define GOSSIP_START_EVENT2     "I am ready for the next challenge."
+#define GOSSIP_START_EVENT_SKIP "I am ready. However I'd like to skip the pageantry."
```
It would be very nice if you can convert these `#define` lines to DB text and use DB references instead of hardcoded text.

Wilcalaf replied:
I dont know yet, but ill research, it doesnt sound that hard

Wilcalaf replied:
Whats the right way to commit DB changes? an SQL file on the commit?

tkrokli replied:
Yes, add SQL files to the same branch as this PR so the content can be checked in travis-ci too.

Treeston replied:
use something like `9999_99_99_world_TOC5.sql` so it's easy to find in the folder

tkrokli commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -15,410 +15,845 @@
+    // Argent Lightwielder
+    SPELL_BLAZING_LIGHT = 67247,
+    SPELL_CLEAVE = 15284,
+    SPELL_UNBALANCING_STRIKE = 67237,
```
Cosmetic/codestyle: align these values (like they used to be in the old code) when your PR gets close to ready for merging (also further down in this file and the other files).

Wilcalaf committed:
Costmetics changes.

tkrokli commented:
This looks pretty neat. What you need now is feedback from other testers.

Wilcalaf committed:
Menu options moved to DB.

Wilcalaf committed:
some tab to space again

Wilcalaf commented:
I moved the stuff to the DB and included the SQL file from the previous branch + some that i added (last 8 lines of the file)

tkrokli commented:
Neat-o, good idea. I had forgotten about Appled's PR content. Let's hope this can be tested soon and get you some feedback.
Just a couple of notes on the SQL:
- move https://github.com/Wilcalaf/TrinityCore/blob/ToC5Adaptation/src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/2018_99_99_21427_world.sql **over to your SQL updates folder,** https://github.com/Wilcalaf/TrinityCore/tree/ToC5Adaptation/sql/updates/world/3.3.5
- cosmetic/codestyle: add `<CR/LF>` to the end of the SQL file

Wilcalaf committed:
SQL moved, added line at end of file

tkrokli commented on sql/updates/world/3.3.5/2018_99_99_21427_world.sql:
```diff
@@ -0,0 +1,448 @@
+(@GOSSIP_MENU,'4','0','[GM] Start Grand Champions encounter, unskipped roleplaying','0','1','1','0','0','0','0','','0','0'),
+(@GOSSIP_MENU,'5','0','[GM] Start Grand Champions encounter, skipped roleplaying','0','1','1','0','0','0','0','','0','0'),
+(@GOSSIP_MENU,'6','0','[GM] Start Eadric the Pure encounter','0','1','1','0','0','0','0','','0','0'),
+(@GOSSIP_MENU,'7','0','[GM] Start Argent Confessor Paletress encounter','0','1','1','0','0','0','0','','0','0'),
```
Please remove the apostrophes from the all the _number_ values (codestyle).

Wilcalaf committed:
SQL apostrophes fixed. Eadric achievement criteria awarded now

Wilcalaf committed:
Sql Typo , the auto compiler is complaining :(

tkrokli commented:
There is a weird error in travis-ci regarding the SQL content of this PR:
```
$ cat sql/updates/world/3.3.5/*.sql | mysql -utrinity -ptrinity world
Warning: Using a password on the command line interface can be insecure.
ERROR 1064 (42000) at line 35546: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '﻿






DELETE FROM `disables` WHERE `entry`=650 AND `sourceType`=2' at line 1


The command "cat sql/updates/world/3.3.5/*.sql | mysql -utrinity -ptrinity world" exited with 1.
```
Line 1-10 of `2018_99_99_21427_world.sql`:
```sql
-- Trial of the Champion (originated from Appled). Thanks to Aokromes for providing sniffs

-- General:

-- Disabling mmap for Trial of the Champion instance
-- creatures are spawned according to sniffs behind an invisible wall and they don't know how to go through it
-- Edit: Aokromes has pushed this already, but disabled the whole instance in the same commit
DELETE FROM `disables` WHERE `entry`=650 AND `sourceType`=2;
-- DELETE FROM `disables` WHERE `sourceType`=7 AND `entry`= 650;
-- INSERT INTO `disables` (`sourceType`,`entry`) VALUES (7, 650);
```
Maybe travis-ci reacts to something in the comments, I am not sure.

Suggestion: try moving the SQL query (DELETE FROM `disables` WHERE `entry`=650 AND `sourceType`=2;) before the comment lines (to Line 1) and see if the error changes or not.

BTW: in case you were wondering, travis-ci only loads an empty TDB from https://github.com/Wilcalaf/TrinityCore/blob/ToC5Adaptation/sql/base/dev/world_database.sql to have something to test the SQL content against, so there could be something caused by the empty DB for the `disables` table (although the DELETE FROM `disables` [...] command should not do anything other than return an empty result (or at least I think it should).

Wilcalaf committed:
Omg i must be tired already lol

Wilcalaf commented:
Mmmm I dont know why im testing this Query on SQLyog and its also complaining about something on first line unless i delete the comments 

> ﻿-- Trial of the Champion (originated from Appled). Thanks to Aokromes for providing sniffs
> 
> -- General:
> 
> -- Disabling mmap for Trial of the Champion instance
> -- creatures are spawned according to sniffs behind an invisible wall and they don't know how to go through it
> -- Edit: Aokromes has pushed this already, but disabled the whole instance in the same commit

I dont know whats going on. but if i delete the comments it will succeed without errors...

Wilcalaf committed:
This one is not complaining after i manually retyped the comment...

Wilcalaf committed:
tkrokli suggestion

tkrokli commented:
Could be a non-ASCII letter in the original text (just guessing).
There are more comment lines than we need anyway.

Wilcalaf committed:
Death knight fly mount movement speed fixed.
No more coffee between until he arrives :)

Wilcalaf commented:
Hey guys - what else i can do to make this more "merge ready" ? thanks!!

Treeston commented:
I'll review in a bit.

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
 
-        SpellScript* GetSpellScript() const override
+    void AttackStart(Unit* who) final override
+    {
```
I _really_ don't like this sort of `AttackStart` replacement. React states instead, please.

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+        switch (id)
         {
-            return new spell_eadric_radiance_SpellScript();
+        case POINT_PREFIGHT:
```
case labels should be indented

```c++
switch (id)
{
    case POINT_PREFIGHT:
        if (whatever)
```

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                switch (eventId)
                 {
-                    if (target->IsAlive())
+                case EVENT_RADIANCE:
```
indentation, see above

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100.0f, true))
                     {
-                        DoCast(target, SPELL_HAMMER_JUSTICE);
+                        DoCast(target, SPELL_HAMMER_JUSTICE, true);
```
why should this be triggered?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                case EVENT_RADIANCE:
+                    DoCastAOE(SPELL_RADIANCE);
+                    Talk(EMOTE_RADIANCE);
+                    events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, urand(20 * IN_MILLISECONDS, 30 * IN_MILLISECONDS));
```
`events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, 20s, 30s);`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                        Talk(EMOTE_HAMMER_RIGHTEOUS, target);
+                        Talk(SAY_HAMMER_RIGHTEOUS);
                     }
+                    events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
```
`std::chrono` please, and explicit urand is unneeded, see above

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                case EVENT_CHALLENGE_DONE:
+                    me->RemoveAllAuras();
+                    me->GetThreatManager().ClearAllThreat();
+                    me->SetFullHealth();
```
Should there be an immunity applied here somewhere? Is it intended that players can still hit him?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
 
-                uiRadianceTimer = 16000;
-            } else uiRadianceTimer -= uiDiff;
+            if (!UpdateVictim())
```
I really dislike `UpdateVictim` needing to be after the event handling loop. Can we change that somehow?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                Talk(SAY_MEMORY_SUMMON);
+                DoCastAOE(SPELL_SHIELD);
+                DoCastAOE(SPELL_CONFESS);
+                me->AttackStop();
```
Should she keep attacking? If not, use `REACT_PASSIVE` here.

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+            argent_challenge_baseAI::JustSummoned(summon);
+            
+            ObjectGuid _memoryGuid = summon->GetGUID();
+            _memory = ObjectAccessor::GetCreature(*me, _memoryGuid);
```
NO - don't do this.

You are not allowed to store any entity pointer between server ticks - units can despawn, or be unloaded for some other reason - then you're dereferencing an invalid pointer, and get some _lovely_ bugs.

Store the GUID and access it using `ObjectAccessor` when you need it - and `nullptr` check the `ObjectAccessor` return!

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                            DoCastSelf(SPELL_RENEW);
+                    }
+
+                    events.Repeat(15 * IN_MILLISECONDS, 17 * IN_MILLISECONDS);
```
`events.Repeat(15s, 17s)`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                case EVENT_HOLY_FIRE:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 30.0f, true))
+                        DoCast(target, SPELL_HOLY_FIRE);
+                    events.Repeat(9 * IN_MILLISECONDS, 12 * IN_MILLISECONDS);
```
`events.Repeat(9s, 12s)`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                case EVENT_HOLY_SMITE_E:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 50.0f, true))
+                        DoCast(target, SPELL_SMITE);
+                    events.Repeat(2 * IN_MILLISECONDS, 3 * IN_MILLISECONDS);
```
`events.Repeat(2s, 3s)`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                        DoCast(target, SPELL_SUMMON_MEMORY, true);
+                    break;
+                case EVENT_MEMORY_AGGRESSIVE:
+                    if(_memory) {
```
code style

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                case EVENT_CHALLENGE_DONE:
+                    me->RemoveAllAuras();
+                    me->GetThreatManager().ClearAllThreat();
+                    me->SetFullHealth();
```
same as eadric - should there be an immune state here?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
-                DoCastAOE(SPELL_CONFESS, false);
+                _shielded = true;
+                damage = me->GetHealth() - 1;
+                me->AttackStop();
```
consider `REACT_PASSIVE`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+
+        void JustDied(Unit* /*killer*/) override
+        {
+            _instance->SetData(DATA_ARGENT_SOLDIER_DEFEATED, _instance->GetData(DATA_ARGENT_SOLDIER_DEFEATED) + 1);
```
Is there any reason to invoke `GetData` here? Is there ever a situation where `DATA_ARGENT_SOLDIER_DEFEATED` is modified, except increments by 1?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+            {
+            case NPC_ARGENT_LIGHWIELDER:
+                _events.ScheduleEvent(EVENT_BLAZING_LIGHT, 10 * IN_MILLISECONDS);
+                _events.ScheduleEvent(EVENT_CLEAVE, urand(4 * IN_MILLISECONDS, 6 * IN_MILLISECONDS));
```
`EVENT_CLEAVE, 4s, 6s`

_(that applies to all eventmap calls, I'll stop mentioning them all individually now)_

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+
+        void AttackStart(Unit* who) override
+        {
+            if (!(me->GetEntry() == NPC_ARGENT_MONK && me->HasAura(SPELL_DIVINE_SHIELD)))
```
ugly! react state passive will take care of all of this for you

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (!UpdateVictim())
```
why is this after the eventmap invocation and casting state check? should npcs not evade while casting?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                switch (eventId)
+                {
+                case EVENT_BLAZING_LIGHT:
+                    if (Unit* friendly = DoSelectLowestHpFriendly(40.0f))
```
should he not heal himself if no friendlies are at/below 40%?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, [this](Unit* u) -> bool { return u->IsWithinDist(me, 5.f) && u->IsNonMeleeSpellCast(true); }))
+                        DoCast(target, SPELL_PUMMEL);
+
+                    _events.Repeat(12 * IN_MILLISECONDS);
```
Should this really be 12s even if we didn't cast it at all?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+        npc_fountain_of_lightAI(Creature* creature) : NullCreatureAI(creature)
         {
-            MemoryGUID = summon->GetGUID();
+            me->SetReactState(REACT_PASSIVE);
```
i think this is part of base `NullCreatureAI` anyway (I could be wrong - check)

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+            me->DespawnOrUnsummon();
+        }
+
+        void AttackStart(Unit* /*who*/) override { }
```
This is not necessary, you're inheriting from `NullCreatureAI` anyway

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -104,321 +150,710 @@ enum Spells
+
+        void AttackStart(Unit* /*who*/) override { }
+
+        void UpdateAI(uint32 /*diff*/) override { }
```
as above

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
 
         void Reset() override
         {
+            _events.Reset();
```
Will there ever be a reasonable `Reset` invocation? If no, why does this still exist? If yes, when?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
-            }else uiShadowPastTimer -= uiDiff;
+            }
+
+            if (!UpdateVictim())
```
why is this after the event handling loop? what event needs to be handled even when not engaged?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
     };
 
-    CreatureAI* GetAI(Creature* creature) const override
+    SpellScript* GetSpellScript() const override
```
all of these could use the generic spell script loader (see fb87ac8e8da3d215d4988b8c2fa25d7a98cf9ee0)

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -644,52 +1086,180 @@ uint32 const memorySpellId[25] =
+        bool Validate(SpellInfo const* /*spellInfo*/) override
         {
-            PrepareSpellScript(spell_paletress_summon_memory_SpellScript);
+            return ValidateSpellInfo({
```
you already have a perfectly good const that has this data, don't replicate

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
+    }
+};
+
+std::vector<uint32> const memorySpells =
```
i would prefer `std::initializer_list` here

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:
```diff
@@ -644,52 +1086,180 @@ uint32 const memorySpellId[25] =
-            }
+        void FilterTargets(std::list<WorldObject*>& targets)
+        {
+            if (targets.empty())
```
`RandomResize` replaces this entire function body

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                if (Creature* knight = instance->GetCreature(DATA_BLACK_KNIGHT))
+                    // If corpse explosion hits any player during encounter
+                    // we cannot give achievement id 3804
+                    knight->AI()->SetData(1, 0);
```
magic number

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+            if (me->HasAura(SPELL_GHOUL_EXPLODE))
+                return;
+            DoCast(me, SPELL_GHOUL_EXPLODE, true);
+            // wtf?
```
i agree with the comment

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+        void JustDied(Unit* /*killer*/) override
         {
-            summons.Despawn(summon);
+            me->DespawnOrUnsummon(1000);
```
`me->DespawnOrUnsummon(1s)`

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (!UpdateVictim())
```
why is this after the eventmap call/casting check? should these not evade while casting?

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
-                            break;
+                            ResetThreatList();
+                            AddThreat(target, 10.0f);
+                            me->AI()->AttackStart(target);
```
`me->AI()` is `this`

also I don't like manual `AttackStart` calls here, `UpdateVictim` will do this anyway

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+        void EnterEvadeMode(EvadeReason why = EVADE_REASON_OTHER) override
+        {
+            BossAI::EnterEvadeMode(why);
+            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_IMMUNE_TO_PC);
```
Don't use flag operations on `IMMUNE_TO_x` flags, use `->SetImmuneToX` instead

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+
+                    // Army of the Dead
+                    // disabling movement temporarily so the spell wont get interrupted
+                    me->SetControlled(true, UNIT_STATE_ROOT);
```
why is this necessary

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
-                switch (uiPhase)
+                // TODO: According to sniffs, The Black Knight should update
+                // creature template to another entry, not just change his display id
+                // But the other templates are missing correct damage modifiers
```
^there's a todo there

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                return;
 
-            instance->SetData(BOSS_BLACK_KNIGHT, DONE);
+            if (id == 0)
```
magic numbers are bad

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                summon->CastSpell(summon, SPELL_DESECRATION_ARM, true);
+                return;
+            }
+            summons.Summon(summon);
```
just forward to `BossAI::JustSummoned` here

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
-    npc_risen_ghoul() : CreatureScript("npc_risen_ghoul") { }
+        void SummonedCreatureDespawn(Creature* summon) override
+        {
+            summons.Despawn(summon);
```
why is this override necessary

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                damage = 0;
+                events.Reset();
+                me->RemoveAllAuras();
+                me->SetHealth(0);
```
ew

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                {
+                    Creature* ghoul = ObjectAccessor::GetCreature(*me, *itr);
+                    if (ghoul && ghoul->IsAlive() && !ghoul->HasAura(SPELL_GHOUL_EXPLODE))
+                        ENSURE_AI(npc_risen_ghoul::npc_risen_ghoulAI, ghoul->AI())->DoExplode();
```
why not just use `DoAction` or `SetData`

Warpten replied:
Actually not a "why not", ENSURE_AI should be avoided at all costs

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+                me->SetHealth(0);
+                me->SetStandState(UNIT_STAND_STATE_DEAD);
+                DoCast(me, SPELL_BLACK_KNIGHT_DIE);
+                for (SummonList::const_iterator itr = summons.begin(); itr != summons.end(); ++itr)
```
why not standard iteration syntax (`for (ObjectGuid const& guid : summons)`)

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
         {
-            uiAttackTimer = 3500;
+            Talk(SAY_DEATH);
+            instance->DoUpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_BE_SPELL_TARGET, SPELL_KILL_CREDIT, 0, me);
```
why this instead of actually casting the spell

Warpten replied:
If not in DBC add it to db table

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -15,337 +15,470 @@
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
             if (!UpdateVictim())
```
same as everywhere - why is this after the update/casting

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -358,7 +491,73 @@ class npc_black_knight_skeletal_gryphon : public CreatureScript
+
+        void SetData(uint32 uiType, uint32 /*uiData*/) override
+        {
+            if (uiType == 1)
```
ew magic numbers

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp:
```diff
@@ -372,7 +571,177 @@ class npc_black_knight_skeletal_gryphon : public CreatureScript
+
+        void RemoveFlag(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+        {
+            // that is wrong, movement disabled by spell (channel)
```
> `// that is wrong`

^i agree with that comment, this should not be necessary if you fix the `_ROOT` hack on spell cast too

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -16,577 +16,1263 @@
-    SPELL_EVISCERATE_H              = 68317,
-    SPELL_FAN_OF_KNIVES             = 67706,
-    SPELL_POISON_BOTTLE             = 67701
+    SPELL_DISENGAGE                     = 68339, // 68340 is correct disengage but spell needs to be scripted
```
i agree with the comment, `68340` needs to be scripted if it's the correct spell...

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -16,577 +16,1263 @@
+    SPELL_MULTI_SHOT                    = 66081,
+    SPELL_SHOOT                         = 66079,
+
+    // Lana Stouthammer && Deathstalker Visceri || Rouge
```
> Rouge

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -16,577 +16,1263 @@
+void EnterAggressiveMode(Creature* temp)
 {
-    Map::PlayerList const& PlList = temp->GetMap()->GetPlayers();
+    temp->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_IMMUNE_TO_PC);
```
immunity flags, `SetImmuneToX`, etc

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -16,577 +16,1263 @@
     }
-
-    return false;
+    DoCastPennant(me);
```
all this pennant logic makes me wonder if all of this couldn't simply be handled by an appropriate entry in `creature_template_addon`'s aura column...

Treeston commented on src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp:
```diff
@@ -16,577 +16,1263 @@
+
+    bool CanUseNormalAI()
+    {
+        // must check for unattackable flag to prevent crash in certain situations
```
"certain situations"?

Treeston reviewed (commented):
didn't have time to review all the files, but this should give you some things to start on

Keader commented:
I wonder if is not better split it in small prs
Review small prs (1 pr per boss) is easy
Review big prs ( rewrite a full instance ) is hard :/

Treeston commented:
i would agree to a merge once the scripts are all in a "workable" state, to be followed up on.

Wilcalaf commented:
Wow - you killed me hahaha. Anyway i gave a brief look to the review and im not sure that ill be able to complete all with my current level. ill give it a shot anyway and try to change as much as i can.
Thanks for review !

Treeston commented:
You'll be fine. Ask if anything is unclear. Worst case you miss up and I tell you to change it again. 😅

Warpten commented on src/server/game/Spells/SpellInfo.cpp:
```diff
@@ -2268,6 +2268,15 @@ void SpellInfo::_LoadSpellDiminishInfo()
                 else if (Id == 51750)
                     return DIMINISHING_NONE;
+				// Triggered trample aura (ToC 5)
+				else if (Id == 67868)
```
Also tabs

Warpten commented on src/server/game/Spells/SpellMgr.cpp:
```diff
@@ -3792,6 +3792,45 @@ void SpellMgr::LoadSpellInfoCorrections()
         spellInfo->Effects[EFFECT_0].Effect = 0;
     });
+	
+	 //
```
Tabs

tkrokli replied:
@Wilcalaf : Depending on the editor you use, look up how to configure it to automatically change inserted tabs to 4 spaces.

Wilcalaf replied:
@tkrokli Aye i fixed it already this file was edited before your reviews - thanks :)

Nitair commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -4678,7 +4678,11 @@ void Spell::EffectLeapBack(SpellEffIndex effIndex)
     //1891: Disengage
-    unitTarget->JumpTo(speedxy, speedz, m_spellInfo->SpellIconID != 1891);
+   //4022: Rolling Throw
+    m_caster->JumpTo(speedxy, speedz, (m_spellInfo->SpellIconID != 1891 && m_spellInfo->SpellIconID != 4022));
```
m_caster should be unitTarget (m_caster becomes also invalid after https://github.com/TrinityCore/TrinityCore/pull/21260)

tkrokli replied:
Thank you for the reminder, good to know.

pedroevil51 commented:
Can this be merged? @Treeston

illfated commented:
@pedroevil51 : not likely, requested changes have not been implemented.
@Wilcalaf : are you going to get back to working on this PR any time soon?

jackpoz commented:
hey @Wilcalaf , do you have time to continue working on this PR ?

jackpoz commented:
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

Aokromes commented:
sugestion, if anyone takeovers this PR split it on smaller parts to get it merged faster.

RR2739-VCN commented:
I am willing to put 100$ bounty on this issue. 

Let's reopen this @Aokromes @jackpoz @Treeston

Keader commented:
It's not a issue, it's a PR.
Issue is still open here: #383

illfated commented:
@RR2739-VCN : To be precise, you can add your bounty to issue #383 (if that was not obvious from Keader's reply).

Aokromes commented:
anyone can takeover this PR?

RR2739-VCN commented:
I am willing to finance the PR. If anyone wants to take over and split it into multiple PRs for a quicker merge it would be great. Msg me if you are taking over this PR

jackpoz commented:
I can host whatever new version of the PR on the TC test realm for a while if anyone is actively developing it. The splitting of the PR is not strictly needed (that was more of a mistake to begin with to create such a huge PR) as long as whoever will pick it up will review ALL the changes and test/debug them properly
